I1073 Control which notifications on the event feed can use collections

johnbrvc opened the pull request:
### Description of what the PR does
Allow specification of when to use notification collections on the event feed in the `pc2v9.ini `file.
Change default setting for the `disable-collections` property in the `pc2v9.ini` file due to the way the ICPC tools CDS works.  We do not want to send `accounts` as collections since the CDS will interpret a collection as a complete list of accounts and remove any accounts configured by the CDS.  This is not desirable since it would loose the CDS' 'admin' account privileges and other CDS accounts.

CI: Give the EventFeed feeder log (list of event feed notifications sent) a unique name based on the client doing the feeding.   This will prevent event feed log corruption if someone decides to run 2 event feeders are run on different ports, for example. Ex.
Old Name: `Eventfeed_2023_06_SumH.log` 
New Name after CI: `Eventfeed_2023_06_SumH_feeder1.log`
CI: change endpoint from "**team-members**" to "**persons**", even though we don't use this at the moment.  We want it right for the future.

### Issue which the PR addresses
Fixes #1073
### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

1. Load and bring up the clics_sumithello sample contest.
2. Generate a feeder account (feeder1).
3. Start the event feeder client using the newly generated feeder1 account.
4. Start the event feed feeding the CLICS 2023-06 API.
5. Using curl, or another utility, access the event feed: https://administrator1:administrator1@localhost:50443/contests/SumH/event-feed
6. Note that the accounts notifications are now individual instead of grouped into a collection (like the other notifications: teams, problems, etc).  
[SampleEventFeed.txt](https://github.com/user-attachments/files/19835577/SampleEventFeed.txt)

7. You can try setting the disable-collections property in the pc2v9.ini file to "all", or add additional notification types to the list to test it out.
8. You can leave out (comment out) the disable-collections line in the pc2v9.ini file to allow collections on all notification types.

johnbrvc committed:
i_1073 Implement notification type collection control

Allow specification of when to use notification collections on the event feed in the pc2v9.ini file.
CI: change endpoint from "team-members" to "persons", even though we don't use this at the moment.

johnbrvc committed:
i_1073 Update default pc2v9.ini for CDS accounts

Due to the way the ICPC tools CDS works, we do not want to send "accounts" as collections since the CDS will interpret a collection as a complete list of accounts and remove any accounts configured by the CDS.  This is not desirable since it would loose the CDS' 'admin' account.
CI: Give the EventFeed feeder log (list of event feed notifications sent) a unique name based on the client doing the feeding.   This will prevent event feed log corruption if someone decides to run 2 event feeders are run on different ports, for example. Ex.  Old Name: Eventfeed_2023_06_SumH.log  New Name after CI: Eventfeed_2023_06_SumH_feeder1.log

SamanwaySadhu reviewed (approved):
LGTM!

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -46,14 +48,25 @@ public class EventFeedJSON extends JSON202306Utilities {
+            String [] disList = CommaSeparatedValueParser.parseLine(IniFile.getValue(CLICS_DISABLE_COLLECTIONS_KEY));
+            this.disableNotificationCollections(disList);
+        } catch (Exception e) {
+            // Use default values for collections
```
It seems like there should at least be a log message generated when this happens.  Otherwise, any error in parsing the IniFile effectively becomes "silently ignore the error".

johnbrvc replied:
Fixed in next push.  This was more work than it was worth, I think....

johnbrvc replied:
I added this log message:
`250425 113341.319|WARNING|qtp1050320876-39|log|Error parsing INI key: clics.disable-collections ("accounts): java.lang.Exception: String missing closing double quote '"accounts'`

clevengr replied:
I didn't really intend that you actually figured out how to tell the user what the specific problem was -- I only suggested that a log message indicating there was problem would be better than "silently ignoring" the problem.  I can see why it was a lot of work doing what you did...

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+
+    /**
+     * Tests if notification collections should be used for a specific notification type.
+     * This first tests is notification collections are being used in general, then if so,
```
I think the word "is" was intended to be "if"?

johnbrvc replied:
Yes.  Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+    }
+
+    /**
+     * Adds the comma-separated list of notification types (EventFeedType's) to the HashSet of
```
I think it's confusing to refer here to "comma-separated list" -- it makes the reader think that what's being passed into this method is a comma-separated list, when in fact it's not -- the "comma-separators" were removed by the call to `CommaSeparatedValueParser.parseLine()` (around line 65).  What's being passed into this method is a String array of notification types which are to be disabled.

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+    }
+
+    /**
+     * Remove the comma-separated list of notification types (EventFeedType's) from the Hashset
```
Same as the previous comment:  I think it's confusing to refer here to "Removing a comma-separated list..." -- there's no comma-separated list being processed by this method.

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+     * notification and will simply set the useCollections flag to true and clear the HashSet of
+     * disabled types.
+     *
+     * @param list Comma-separated list of notification types (EventFeedType's) to allow collections on.
```
As previously mentioned, referring to a "comma-separated list" here is confusing, because that's not what the method is receiving.

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+     * on the event feed.  The special type ALL_NOTIFICATION_TYPES will disable collections for all
+     * notification and will clear the HashSet of disabled types and simply set the useCollections flag to false.
+     *
+     * @param list Comma-separated list of notification types (EventFeedType's) to not use
```
As mentioned above, the actual parameter here is not a comma-separated list; I think this is confusing.

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+
+    /**
+     * Remove the comma-separated list of notification types (EventFeedType's) from the Hashset
+     * of disabled notification types.  If collections are enabled, then the types included in the
+     * supplied list will be allowed to generate collections.  The special type ALL_NOTIFICATION_TYPES will disable collections for all
```
The above comment reads wrong to me for a couple of reasons.  

First, it says `if collections are enabled then the types included in the supplied list will be allowed to generate collections`.  If I'm reading the following code correctly, then when `ALL_NOTIFICAITON_TYPES` is specified, the effect is that _all_ notifications are allowed to use collections -- not just "the types included in the supplied list".  (Perhaps there is some confusion between the definitions "collections are enabled" and "ALL_NOTIFICATION_TYPES" -- but in any case I can't reconcile the comment with what it appears the code is actually doing.  But since there's actually no test in the method code for the condition "collections are enabled", I'm assuming that's what the test `if(notificationType.equals(ALL_NOTIFICATION_TYPES))` is testing.  But again, I can't reconcile the difference between what the comment says and what the code appears to be doing.)

Second, the comment says `The special type ALL_NOTIFICATION_TYPES will _disable_ [emphasis mine] collections for all notification...`.  It appears to me that what the code is actually doing in this case is _enabling_ all notifications, not disabling them.

johnbrvc replied:
Maybe the confusion is, "If collections are enabled, then..." The intention here is: when making use of these flags and the Hashset, if the "useCollections" member is true, meaning, that collections are enabled, eg "isUseCollections()", then we are effectively removing these items in the supplied list from the hashset, thereby allowing those things in the array to generate collections.  So, it "undoes" the disableNotificationCollections for a list of notification types, allowing them to generate collections, iff useCollections is true.

Setting *any* element of the supplied list[] to ALL_NOTIFICATION_TYPES (all) basically enables everything to generate collections, regardless of any other things on the list.  "all" overrides everything in both cases (disable and enable).

clevengr commented on src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java:
```diff
@@ -1056,13 +1069,80 @@ public void setFilter(Filter filter) {
+     *
+     * @param list Comma-separated list of notification types (EventFeedType's) to allow collections on.
+     */
+    public void enableNotificationCollections(String [] list) {
```
It does not appear that this method is invoked anywhere in the workspace (pc2v9 project).  Is that intentional?

johnbrvc replied:
Yes, since I added a "disableNotificationCollections", it seemed fitting to add an "enableNotificationCollections" for future use.

clevengr reviewed (changes requested):
Reviewed changes to `src/edu/csus/ecs/pc2/clics/API202306/EventFeedJSON.java`.  Request one change (addition of logging of an error condition), plus recommend several changes to documentation.

clevengr reviewed (commented):
Reviewed all the changed files.  Requested one change in `EventFeedJSON`, plus several recomended (minor) fixes.  All the other files look good.  Proceeding to testing.

clevengr commented:
Completed execution test; appears to work as advertised.  I'm ready to approve pending response to my comment(s) in `EventFeedJSON`.

johnbrvc committed:
i_1073 PR changes requests

Add a log message if the disable collections key is improperly formatted.
Clean up some comments.

clevengr reviewed (approved):
Looks good; I approve this PR.
